### DISCUSSION TITLE
How to build fish man pages from source

### DISCUSSION THREAD

**Discussion by dtenenba** [fish-shell/fish-shell#10897] [General]
Hi, I have some funky older systems where I have to build fish from source (because reasons).

`cargo build --release` works perfectly. However, it gives this warning:

```
warning: fish@4.0.0-alpha1: Could not build man pages: Os { code: 2, kind: NotFound, message: "No such file or directory" }
```

I am not sure what file or directory it is complaining about. Adding `-vv` to the `cargo build` command did not shed light.  I made sure pandoc, help2man, and sphinx-build are available but the error still happens. 

To be clear, fish is built and runs, but no fish-specific man pages (for e.g. `set` and `syntax`) are available, whereas they are on systems where I have installed fish from a package manager.

How do I find out what that error is about?

This is with c58313fb2bd0a53795d61b0a26530f52ed2474d9 on Ubuntu 18.04, cargo 1.83.0 (5ffbef321 2024-10-29), rustc 1.83.0 (90b35a623 2024-11-26).

**Comment by dtenenba** [comment]
Removing the `target` directory and re-building (after making `sphinx-build` available) somehow fixed this.

**Comment by dtenenba** [comment]
But there are missing man pages compared to other systems. 

The following commands fail on the system described above but work on systems where fish was installed via a package manager:

```
fg -h
man syntax
```

How can I build all the goodies?

**Comment by zanchey** [comment]
Have you used the self-installable instructions, or the CMake instructions? The CMake instructions are probably your best bet.

**Comment by dtenenba** [comment]
I used `cargo build`. I had some trouble with both `cargo install` and the CMake instructions.

**Comment by dtenenba** [comment]
I ended up using the CMake method and now I have most of the man pages. Thanks.

**Comment by faho** [comment]
> I made sure pandoc, help2man, and sphinx-build are available but the error still happens.

We only need sphinx-build. Neither pandoc nor help2man are involved.

> no fish-specific man pages (for e.g. set and syntax) are available

There is no "syntax" man page. There *is* a "fish-language" page.

> warning: fish@4.0.0-alpha1: Could not build man pages: Os { code: 2, kind: NotFound, message: "No such file or directory" }

This most likely means that it could not find sphinx-build.

I'm going to uncouple the errors here a bit, because *technically* there are other files it could be missing, but 98% it's sphinx-build.

Cargo appears to be doing some caching here, I think if you install sphinx after building once you need to clear the cache. We might be able to get it to re-run if that happens.

What you want to check is that `count ~/.local/share/fish/install/man/man1/*` prints (currently) 132.

**Reply by faho** [reply]
Okay, I think I know what's going on:

You didn't have sphinx installed and ran `cargo install`. That skipped the man pages but printed a warning.

Then you installed sphinx and ran it again, but cargo did not rerun the build script because it wasn't told to.

I don't believe we *can* tell it to "rerun if this command now exists", so the best solution is probably to panic if sphinx isn't installed and let you disable the man pages with an explicit environment variable (`BUILD_DOCS=0 cargo install`).

**Reply by dtenenba** [reply]
> What you want to check is that `count ~/.local/share/fish/install/man/man1/*` prints (currently) 132.

Oddly this prints 131 for me. Close enough.

**Reply by faho** [reply]
Okay, 2d2f18c159de890ab17dd65bc2f7f8d53d6c5b3f will now make the cargo build panic if sphinx isn't installed, unless you set FISH_BUILD_DOCS=0.

This will mean if you install sphinx it will rerun the build script, rebuilding the man pages.

This isn't *ideal* - I would love if we could skip the man page generation, but I do not know how to tell cargo to rerun build.rs if $PATH/sphinx-build changed.
